File: fun2/fund2-src/src/c/v1/main_app.c
#include "account_v1.h"

int main() 
{
    Account *acc = create_account("1234-0001", 100.);

    // Illegal argument 
    // Account *acc = create_account("1234-0001", -100.);

    // Unavoidable invalid creation
    /*
    Account *acc = (Account *) malloc(sizeof(Account));
    acc->account_number = "1234-0002";
    acc->balance = -200;
    */
    printf("Account %s has been created with a balance of %.2f baht.\n",
        get_account_number(acc), get_balance(acc));

    deposit(acc, 500.);
    withdraw(acc, 400.);

    // overdrawn balance
    withdraw(acc, 800.);

    // unauthorized direct update
    // acc->account_number = "1234-0002";
    printf("Account %s has been created with a balance of %.2f baht.\n",
        get_account_number(acc), get_balance(acc));

    delete_account(acc);
    // Illegal access to dangling pointer
    printf("Deleted account %s had a balance of %.2f baht.\n", 
        get_account_number(acc), get_balance(acc));
    acc = NULL;
    // Access to NULL Pointer
    /*printf("Deleted account %s had a balance of %.2f baht.\n", 
        get_account_number(acc), get_balance(acc));*/
    return 0;
}

File: fun2/fund2-src/src/main/java/simplebank/v3/MainApp.java
package simplebank.v3;

public class MainApp {
    public static void main(String[] args) {
        Account acc = new Account("1234-0001", 100.);
        System.out.printf(
            "Account %s has been created with a balance of %.2f baht.\n",
            acc.getAccountNumber(), acc.getBalance());
        try {
            acc.deposit(500.);
            acc.withdraw(400.);
            acc.deposit(-800);
            acc.withdraw(800.);
        } catch (IllegalArgumentException e) {
            System.out.printf("Illegal operation: %s\n", e.getMessage());
        }
        System.out.printf("Account %s now has a balance of %.2f baht.\n", 
            acc.getAccountNumber(), acc.getBalance());
        acc = null;
    }
}
